Add 2D RoPE

csjfwang opened the pull request:
## Description

Add 2D RoPE to Global and Forecast Engine (on all forecast steps). 

<!--
Provide a brief summary of the changes introduced in this pull request.

Change the title of the PR to a short sentence easy to understand:
[1234][model] Adds FSDP2 monitoring code
-->


## Issue Number

<!--
Link the Issue number this change addresses:
Closes #XYZ 
-->

Closes #1109 

## Checklist before asking for review

-   [x] I have performed a self-review of my code
-   [x] My changes comply with basic sanity checks:
      - I have fixed formatting issues with `./scripts/actions.sh lint`
      - I have run unit tests with `./scripts/actions.sh unit-test`
      - I have documented my code and I have updated the docstrings.
      - I have added unit tests, if relevant
-   [x] I have tried my changes with data and code:
      - I have run the integration tests with `./scripts/actions.sh integration-test`
      - (bigger changes) I have run a full training and I have written in the comment the run_id(s): `launch-slurm.py --time 60`
      - (bigger changes and experiments) I have shared a hegdedoc in the github issue with all the configurations and runs for this experiments
-   [ ] I have informed and aligned with people impacted by my change:
    - for config changes: the MatterMost channels and/or a design doc
    - for changes of dependencies: the MatterMost software development channel

Commit:
Replace cf.rank==0 with utils.distributed.is_root

Commit:
replace cf.rank==0 with weathergen.utils.distributed.is_root

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

Commit:
add 2d rope to develop

Commit:
simplify assimilate global, forecast mode config

Commit:
add 2d rope to forecast eigine only once

Commit:
only keep global & forecast engine add 2d rope

Commit:
simplify the code

Commit:
fix lint

Commit:
small fix

Commit:
fix annotation

Commit:
fix lint

Commit:
add annotation

csjfwang committed:
Merge branch 'ecmwf:develop' into develop

Commit:
Merge branch 'develop' into develop_2d_rope_final_gf

csjfwang commented:
Some experiments here: 
https://gitlab.jsc.fz-juelich.de/hedgedoc/lnorUrxoRC6O0J-rOIS1Mg?view
https://gitlab.jsc.fz-juelich.de/hedgedoc/PtWqwrTHSayrtzJmgFNrQQ?view

Commit:
default config

Commit:
fix default use_reentrant

Commit:
use_2d_rope false as defaut

clessig commented on src/weathergen/model/model.py:
```diff
@@ -83,6 +84,21 @@ def __init__(self, cf) -> None:
         )
         self.pe_global = torch.nn.Parameter(pe, requires_grad=False)
 
+        ### ROPE COORDS (for 2D RoPE when use_2D_rope=True) ###
```
standard formatting for comment, no need for `###` or all caps

clessig commented on src/weathergen/model/model.py:
```diff
@@ -83,6 +84,21 @@ def __init__(self, cf) -> None:
         self.pe_global = torch.nn.Parameter(pe, requires_grad=False)
 
+        ### ROPE COORDS (for 2D RoPE when use_2D_rope=True) ###
+        self.use_2D_rope = cf.use_2D_rope
```
Use `cf.get( 'use_rope_2D`, False)` for backward compatability

clessig replied:
Parameter here and config flag should be `rope_2D` or `use_rope_2D`. That's easier to parse when we have different rope variants potentially in the future

clessig commented on src/weathergen/model/model.py:
```diff
@@ -83,6 +84,21 @@ def __init__(self, cf) -> None:
+        else:
+            self.rope_coords = None
+
         ### HEALPIX NEIGHBOURS ###
```
Standard formatting for comment

clessig commented on src/weathergen/model/model.py:
```diff
@@ -150,28 +167,55 @@ def reset_parameters(self, cf: Config) -> "ModelParams":
+            self.rope_coords.data.copy_(coords_flat)
+
+            # Clear pe_global when using 2D RoPE
+            self.pe_global.data.fill_(0.0)
```
Is this at all needed with 2D rope

csjfwang replied:
Yes, because I kept the following code unchanged:

Line 738 - 743:
```
if self.cf.ae_local_queries_per_cell:
            tokens_global = (self.q_cells + model_params.pe_global).repeat(batch_size, 1, 1)
        else:
            tokens_global = (
                self.q_cells.repeat(self.num_healpix_cells, 1, 1) + model_params.pe_global
            )
```

Line 840 - 844:
```
# recover batch dimension and build global token list
        tokens_global = (
            tokens_global.reshape([batch_size, self.num_healpix_cells, s[-2], s[-1]])
            + model_params.pe_global
        ).flatten(1, 2)
```

clessig commented on src/weathergen/model/positional_encoding.py:
```diff
@@ -94,3 +94,80 @@ def positional_encoding_harmonic_coord(x, lats, lons):
+
+
+####################################################################################################
+def apply_rotary_pos_emb_2d(q, k, coords, base=10000.0, unsqueeze_dim=1):
```
Remove `apply_` from name

clessig commented on config/default_config.yml:
```diff
@@ -63,6 +63,11 @@ impute_latent_noise_std: 0.0  # 1e-4
+# Use 2D RoPE instead of traditional global positional encoding
+# When True: uses 2D RoPE based on healpix cell coordinates (lat/lon)
+# When False: uses traditional pe_global positional encoding
+use_2D_rope: False
```
use_rope_2D: False (see below for explanation)

clessig commented on src/weathergen/model/model.py:
```diff
@@ -636,7 +679,9 @@ def forward(self, model_params: ModelParams, batch, forecast_offset: int, foreca
 
-            tokens = self.forecast(model_params, tokens, fstep)
+            # Apply 2D RoPE coords only on the first forecast step
+            is_first_forecast = fstep == forecast_offset
```
You can just have `fstep == forecast_offset` directly in the function argument

clessig commented on src/weathergen/model/positional_encoding.py:
```diff
@@ -94,3 +94,80 @@ def positional_encoding_harmonic_coord(x, lats, lons):
+
+
+####################################################################################################
+# Rotary positional embeddings (2D) adapted from Qwen3 & LLama for reuse in WeatherGenerator.
```
Can you add to the copyright notice of the file. See Sophie's branch for examples.

clessig reviewed (commented):
Some minor things to address. I think it would be worthwhile (since it's easy, and scientifically plausible) to try out 2D rope in every layer in the global and forecast engines. We should also think how to make sure the decoder gets consistent data, i.e. with a positional encoding that does not change with the forecast step (we also didn't think about this for the current global positional encoding).

Commit:
Add copyright notice for RoPE functions and update naming

- Add copyright attribution for rotate_half() and apply_rotary_pos_emb()
  functions

- Rename apply_rotary_pos_emb_2d() to rotary_pos_emb_2d() for consistency

- Rename config parameter use_2D_rope to rope_2D for better extensibility
  when supporting different RoPE variants in the future

Commit:
fix lint

Commit:
fix lint

Commit:
add 2d rope to all forecast steps

clessig commented:
Will be merged as #1540
